system/core/tags.php

```diff
@@ -0,0 +1,192 @@
+			document.getElementById('tagList').innerHTML = 'loading...';
+		}
+	}
+	function deselectTag(id) {
+		hypha(<?=json_encode($url)?>, 'tagDeselect', id, document.getElementById('<?=$formId?>'), function(response) {document.getElementById('tagList').innerHTML = response;});
+		document.getElementById('tagList').innerHTML = 'loading...';
+	}
+	function addTag() {
+		var name = prompt(<?=json_encode(__('enter-label'))?>, '');
+		if (name) {
+			var description = prompt(<?=json_encode(__('enter-description'))?>, '');
+			if (description) {
+				var json = '{ "name" : "' + name + '", "description" : "' + description + '"}';
+				hypha(<?=json_encode($url)?>, 'tagNew', json, document.getElementById('<?=$formId?>'), function(response) {document.getElementById('tagList').innerHTML = response;});
+				document.getElementById('tagList').innerHTML = 'loading...';
+			}
+		}
+	}
+</script>
+<?php
+		$hyphaHtml->writeScript(ob_get_clean());
+
+		return tagList($pageListNode, $lang);
+	}
+
+	registerCommandCallback('tagSelect', 'tagSelectCallback');
+	registerCommandCallback('tagDeselect', 'tagDeselectCallback');
+	registerCommandCallback('tagNew', 'tagNewCallback');
+
+	function tagSelectCallback($id) {
+		global $hyphaXml;
+		global $O_O;
+
+		$lang = $O_O->getContentLanguage();
+
+		$hyphaXml->lockAndReload();
+		$pageListNode = hypha_getPage($lang, $O_O->getRequest()->getPageName());
+		$newtag = $hyphaXml->createElement('tag');
+		$newtag->setAttribute('id', $id);
+		$pageListNode->appendChild($newtag);
+		$hyphaXml->saveAndUnlock();
+
+		echo tagList($pageListNode, $lang);
+		exit;
+	}
+
+	function tagDeselectCallback($id) {
+		global $hyphaXml;
+		global $O_O;
+		$lang = $O_O->getContentLanguage();
+
+		$hyphaXml->lockAndReload();
+		$pageListNode = hypha_getPage($lang, $O_O->getRequest()->getPageName());
+		foreach($pageListNode->getElementsByTagName('tag') as $tag) if ($tag->getAttribute('id')==$id) $pageListNode->removeChild($tag);
+		$hyphaXml->saveAndUnlock();
+
+		echo tagList($pageListNode, $lang);
+		exit;
+	}
+
+	function tagNewCallback($json) {
+		global $hyphaXml;
+		global $O_O;
+
+		$lang = $O_O->getContentLanguage();
+		$pageListNode = hypha_getPage($lang, $O_O->getRequest()->getPageName());
+
+		$json = json_decode($json, true);
+
+		// Check if tagList exists
+		$tagList = $hyphaXml->findXPath('hypha/tagList');
+		if ($tagList->count() === 0) {
+			$hyphaXml->lockAndReload();
+			$tagList = $hyphaXml->createElement('tagList');
+			$hyphaXml->documentElement->appendChild($tagList);
+			$hyphaXml->saveAndUnlock();
+		}
+
+		// Check if tag exists
+		$path = 'hypha/tagList/tag/language[@id=' . xpath_encode($lang) . ' and @label=' . xpath_encode($json['name']) . ']';
+		$tagExists = $hyphaXml->findXPath($path)->count() > 0;
+
+		// Create new tag, add to tagList and page in hypha.xml
+		if (!$tagExists) {
+			$newTagName = $json['name'];
+			$newTagDescription = $json['description'];
+
+			$hyphaXml->lockAndReload();
```

Comment:
This lock should probably happen before checking whether the tag exists (probably even before checking whether the tag list exists), since otherwise there is a race condition where the tag list or tag might be added twice if two requests happen at the same time. This does mean the file is locked for a bit longer, but that should be fine.